src/main/java/com/artipie/asto/s3/S3Storage.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.s3;
+
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.Transaction;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
+
+/**
+ * Storage that holds data in S3 storage.
+ *
+ * @since 0.15
+ */
+public final class S3Storage implements Storage {
+
+    /**
+     * S3 client.
+     */
+    private final S3AsyncClient client;
+
+    /**
+     * Bucket name.
+     */
+    private final String bucket;
+
+    /**
+     * Ctor.
+     *
+     * @param client S3 client.
+     * @param bucket Bucket name.
+     */
+    public S3Storage(final S3AsyncClient client, final String bucket) {
+        this.client = client;
+        this.bucket = bucket;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Publisher<ByteBuffer> content) {
+        return this.client.createMultipartUpload(
+            CreateMultipartUploadRequest.builder()
+                .bucket(this.bucket)
+                .key(key.string())
+                .build()
+        ).thenApply(
+            created -> new MultipartUpload(this.client, this.bucket, key, created.uploadId())
+        ).thenCompose(
+            upload -> upload.upload(content)
+                .thenCompose(ignored -> upload.complete())
+        ).thenApply(response -> null);
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Publisher<ByteBuffer>> value(final Key key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Transaction> transaction(final List<Key> keys) {
+        throw new UnsupportedOperationException();
```

Comment:
@olegmoz Same for all unsupported operations: I suggest to add a puzzle to support what can be supported in the future.